### ISSUE TITLE
Support openapi/jsonschema value constraints

### ISSUE THREAD

**Issue by adamw** [softwaremill/tapir#48]
When interpreted as openapi docs, this would translate to the following properties from the schema object (https://swagger.io/specification/#schemaObject):
* enum
* minimum, maximum
* pattern
* maxLength/minLength

The `tapir.Schema` trait should be extended with a description of value constraints (default no constraints). 

The encouraged (and documented) way to define constraints would be by creating new types. This could be through e.g. tagged types (which have no runtime overhead, e.g. https://github.com/softwaremill/scala-common#tagging), value types or "normal" custom types.

We should provide a convenient way to define constraints for a type - through builder methods on schema. Possible syntax:

```scala
implicit val schemaForEmail: SchemaFor[String @@ Email] = 
  SchemaFor[String].constraint(Pattern(emailRegex))

implicit val schemaForColors: SchemaFor[String @@ Color] = 
  SchemaFor[String].constraint(Enum("red", "pink", "violet"))
```

The constraints defined on the schema should also be checked during decoding - resulting in a new `DecodeResult.InvalidValue` (or maybe reuse/replace `Mismatch`)?

**Comment by tayvs** [comment]
@adamw @ghostbuster91 Is there any updates?

**Comment by ghostbuster91** [comment]
@tayvs yes, we are working on it. Should be ready in the nearest future, stay tuned :)

**Comment by danxmoran** [comment]
@ghostbuster91 sorry for repeated pings, but are there new updates on this? This feature would be a big convenience for my team, so I'd be happy to help out with finishing it if there are pieces that could be split out.

**Comment by ghostbuster91** [comment]
No problem, I also think that this feature will be very convenient for everyone and would love to ship it. Luckily, recently I made almost complete implementation. It is still not merge-able since I was mostly experimenting and doing some shortcuts but I believe that it is something on which we can iterate. It would be great if you can take a look at #201 and give some feedback :)

**Comment by luksow** [comment]
As discussed here: https://gitter.im/softwaremill/tapir?at=5d5fb1d768406739f96b3712 I'm adding my use case. I'd like, on errors, to return the same `jsonBody[JsonError]` for various errors but different status codes to be a good HTTP/REST citizen. Ex:
```scala
    endpoint
      .post
      .in("auth" / "session")
      .in(jsonBody[LoginRequest])
      .out(oneOf(statusMapping(Created, jsonBody[LoginResponse.LoggedIn])))
      .errorOut(...)
      .serverLogic { request =>
        authService.login(request).map {
          case logged @ LoginResponse.LoggedIn(_, _) => Right(logged)
          case LoginResponse.InvalidCredentials => Left((Unauthorized, JsonError("Invalid credentials"))) // just an idea
          case LoginResponse.UserNotFound => Left((NotFound, JsonError("User not found"))) // just an idea
        }
      }
```

**Comment by adamw** [comment]
Released in 0.10.0
@luksow your comment moved to #208 :)
